src/geometry/regionbasedadjacency.h

```diff
@@ -30,46 +33,28 @@ namespace LibGeoDecomp {
 class RegionBasedAdjacency
 {
 public:
+    RegionBasedAdjacency();
+
     /**
      * Insert a single edge (from, to) to the graph
      */
-    inline void insert(int from, int to) {
-        region << Coord<2>(to, from);
-    }
+    void insert(int from, int to);
 
     /**
      * Insert all pairs (from, x) with x \in to, to the graph. We sort
      * parameter to to ensure fast, linear inserts.
      */
-    inline void insert(int from, std::vector<int> to) {
-        std::sort(to.begin(), to.end());
-        Region<2> buf;
-        for (std::vector<int>::const_iterator i = to.begin(); i != to.end(); ++i) {
-            buf << Coord<2>(*i, from);
-        }
-
-        region += buf;
-    }
+    void insert(int from, std::vector<int> to);
 
     /**
      * Returns all x \in V with (node, x) \in E.
      */
-    inline void getNeighbors(int node, std::vector<int> *neighbors) {
-        CoordBox<2> box = region.boundingBox();
-        int minX = box.origin.x();
-        int maxX = minX + box.dimensions.x();
+    void getNeighbors(int node, std::vector<int> *neighbors) const;
 
-        for (Region<2>::StreakIterator i = region.streakIteratorOnOrAfter(Coord<2>(minX, node));
-             i != region.streakIteratorOnOrAfter(Coord<2>(maxX, node));
-             ++i) {
-            for (int j = i->origin.x(); j < i->endX; ++j) {
-                (*neighbors) << j;
-            }
-        }
-    }
+    const Region<2> &getRegion() const {return *region;}
 
 private:
-    Region<2> region;
+    boost::shared_ptr<Region<2> > region;
```

Comment:
RegionBasedAdjacency includes and requires Region, Region includes and requires RegionBasedAdjacency. One has to use a declaration of the other, thus needs a pointer and cannot use the complete type. I chose RegionBasedAdjacency to hold a pointer, which I prefer to be a shared_ptr.